### ISSUE TITLE
Configuration options for CA1724

### ISSUE THREAD

**Issue by MisinformedDNA** [dotnet/roslyn-analyzers#2522]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.2

#### Diagnostic ID

[CA1724](https://docs.microsoft.com/visualstudio/code-quality/ca1724-type-names-should-not-match-namespaces)

#### Issue

This was once defined (and still is [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.FxCopAnalyzers/Microsoft.CodeAnalysis.FxCopAnalyzers.md)) as

> Type names should not match the names of namespaces that are defined in the .NET Framework class library. Violating this rule can reduce the usability of the library.

Due to the way it was implemented, the description has changed to:

> User-created type names should not match the names of referenced namespaces that have externally visible types. Violating this rule can reduce the usability of your library.

Per https://github.com/dotnet/roslyn-analyzers/issues/1753, this can not be undone. So I was wondering if this can be made configurable in .editorconfig?

#### Example

The specific scenario I ran into was that I was referencing a NuGet package that referenced `AngleSharp.Extensions`. And I had a type named, obviously, `Extensions`. Because it's just a nested reference that I don't expose, so no consumer of my project would ever get confused as AngleSharp.Extensions is never exposed to them.

**Comment by Evangelink** [comment]
@mavasani Is it enough to add the `excluded_symbol_names` option for this rule or are you thinking about something more specific?

**Comment by Evangelink** [comment]
Ping @mavasani

**Comment by Evangelink** [comment]
I am not really sure what's the best solution here. I understand your problem and to be honest I am facing the same kind of issues with some namespaces from Telerik.

@mavasani do you have any opinion?

**Comment by mavasani** [comment]
I am not sure either - `excluded_symbol_names` doesn't really add much value as users can add a source suppression on each symbol.

**Comment by MisinformedDNA** [comment]
Ideas:
1. Only look at namespaces of directly referenced assemblies
2. Create a rule with a list of namespaces to ignore (i.e. "AngleSharp*;Teleric.Components*")

**Comment by Evangelink** [comment]
@MisinformedDNA Could you make a new test with the latest version of the analyzer? I don't know if something changed or if my repro is invalid but I don't manage to see the issue.

**Comment by MisinformedDNA** [comment]
I can't find it right now. I'll reopen if it turns up again.

**Comment by Bouke** [comment]
I would be interested in the exclusion list discussed here. I'm using the Twilio nuget package which contains various generic namespaces with a singular name, like `User`. Now I'm faced with warnings like the following: 

>  CA1724 - The type name User conflicts in whole or in part with the namespace name 'Twilio.Rest.Chat.V1.Service.User'. Change either name to eliminate the conflict.

Allowing me to ignore `Twilio.*` namespaces would be very helpful.

**Comment by MisinformedDNA** [comment]
@Bouke Could you create a small repro of the issue?

**Comment by Bouke** [comment]
@MisinformedDNA sure!

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Twilio" Version="5.80.1" />
  </ItemGroup>
</Project>
```
```cs
namespace CA1724;
public class User { }
```

```
> dotnet build
MSBuild version 17.3.1+2badb37d1 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
CSC : warning CA1014: Mark assemblies with CLSCompliant [/Users/bouke/Developer/CA1724/CA1724.csproj]
/Users/bouke/Developer/CA1724/User.cs(2,14): warning CA1724: The type name User conflicts in whole or in part with the namespace name 'Twilio.Rest.Chat.V1.Service.User'. Change either name to eliminate the conflict. [/Users/bouke/Developer/CA1724/CA1724.csproj]
  CA1724 -> /Users/bouke/Developer/CA1724/bin/Debug/net6.0/CA1724.dll

Build succeeded.
```

**Comment by starteleport** [comment]
Hi!

This issue is relevant for me, too. It would be great if I had a possibility to exclude third-party libraries' namespaces from consideration.

Are there any plans to address this?
